rocketpy/motors/Tank.py

```diff
@@ -0,0 +1,524 @@
+        self.initial_gas_mass = initial_gas_mass
+
+        self.liquid_mass_flow_rate_in = Function(
+            liquid_mass_flow_rate_in,
+            inputs="Time",
+            outputs="Mass Flow Rate",
+            interpolation="linear",
+            extrapolation="zero",
+        )
+        self.gas_mass_flow_rate_in = Function(
+            gas_mass_flow_rate_in,
+            inputs="Time",
+            outputs="Mass Flow Rate",
+            interpolation="linear",
+            extrapolation="zero",
+        )
+        self.liquid_mass_flow_rate_out = Function(
+            liquid_mass_flow_rate_out,
+            inputs="Time",
+            outputs="Mass Flow Rate",
+            interpolation="linear",
+            extrapolation="zero",
+        )
+        self.gas_mass_flow_rate_out = Function(
+            gas_mass_flow_rate_out,
+            inputs="Time",
+            outputs="Mass Flow Rate",
+            interpolation="linear",
+            extrapolation="zero",
+        )
+
+    @funcify_method("Time (s)", "mass (kg)")
+    def mass(self, t):
+        return self.liquidMass(t) + self.gasMass(t)
+
+    @funcify_method("Time (s)", "mass (kg)")
+    @except_negative
+    def liquidMass(self, t):
+        liquid_flow = self.netLiquidFlowRate.integral(0, t)
+        return self.initial_liquid_mass + liquid_flow
+
+    @funcify_method("Time (s)", "mass (kg)")
+    @except_negative
+    def gasMass(self, t):
+        gas_flow = self.netGasFlowRate.integral(0, t)
+        return self.initial_gas_mass + gas_flow
+
+    @funcify_method("Time (s)", "liquid mass flow rate (kg/s)", extrapolation="zero")
+    def netLiquidFlowRate(self):
+        return self.liquid_mass_flow_rate_in - self.liquid_mass_flow_rate_out
+
+    @funcify_method("Time (s)", "gas mass flow rate (kg/s)", extrapolation="zero")
+    def netGasFlowRate(self):
+        return self.gas_mass_flow_rate_in - self.gas_mass_flow_rate_out
+
+    @funcify_method("Time (s)", "mass flow rate (kg/s)", extrapolation="zero")
+    def netMassFlowRate(self):
+        return self.netLiquidFlowRate + self.netGasFlowRate
+
+    @funcify_method("Time (s)", "volume (m³)")
+    def liquidVolume(self):
+        return self.liquidMass / self.liquid.density
+
+    @funcify_method("Time (s)", "volume (m³)")
+    def gasVolume(self):
+        return self.gasMass / self.gas.density
+
+    @funcify_method("Time (s)", "height (m)")
+    def liquidHeight(self):
+        return self.geometry.inverse_volume.compose(self.liquidVolume)
+
+    @funcify_method("Time (s)", "height (m)")
+    def gasHeight(self, t):
+        fluid_volume = self.gasVolume + self.liquidVolume
+        gasHeight = self.geometry.inverse_volume(fluid_volume(t))
+        if gasHeight <= self.geometry.top:
+            return gasHeight
+        else:
+            raise ValueError(
+                f"The tank {self.name}, is overfilled"
+                f"with gas height {gasHeight} at time {t}"
+            )
```

Comment:
This overfilling check needs to be done in `liquidHeight` too.

Reply:
duplicated